[Flaky] Test.DeCov is Flaky

[Jenkins](https://jenkins.preferred.jp/job/chainer/job/chainer_pr/2658/TEST=chainer-py35,label=mn1-p100/testReport/junit/tests.chainer_tests.functions_tests.loss_tests.test_decov.TestDeCov_param_1_%7Bbackward_options=%7B'atol'_%200.03%7D,%20dtype=float16,%20forward_options=%7B'rtol'_%200.01,%20'atol'_%200/01%7D,%20reduce='no'%7D/test_backward_gpu/)
```
tests.chainer_tests.functions_tests.loss_tests.test_decov.TestDeCov_param_1_{backward_options={'atol': 0.03}, dtype=float16, forward_options={'rtol': 0.01, 'atol': 0.01}, reduce='no'}.test_backward_gpu 
```
```
self = <chainer.testing._bundle.TestDeCov_param_1_{backward_options={'atol': 0.03}, dtype=float16, forward_options={'rtol': 0.01, 'atol': 0.01}, reduce='no'} testMethod=test_backward_gpu>

    @attr.gpu
    def test_backward_gpu(self):
        self.check_backward(cuda.to_gpu(self.h),
>                           cuda.to_gpu(self.gloss))

self       = <chainer.testing._bundle.TestDeCov_param_1_{backward_options={'atol': 0.03}, dtype=float16, forward_options={'rtol': 0.01, 'atol': 0.01}, reduce='no'} testMethod=test_backward_gpu>

tests/chainer_tests/functions_tests/loss_tests/test_decov.py:106: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/chainer_tests/functions_tests/loss_tests/test_decov.py:83: in check_backward
    f, (h_data,), gloss_data, eps=0.02, **self.backward_options)
chainer/gradient_check.py:893: in check_backward
    detect_nondifferentiable, is_immutable_params=False
chainer/gradient_check.py:466: in run
    self._run()
chainer/gradient_check.py:509: in _run
    self._compare_gradients(gx_numeric, gx_backward, directions)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <chainer.gradient_check._CheckBackward object at 0x7f83d16deda0>
gx_numeric = array(0.049424171447753906)
gx_backward = array(0.08145464773508873)
directions = [array([[-0.1604053 ,  0.12723993,  0.4412627 ],
       [ 0.5194434 ,  0.30622899,  0.0786798 ],
       [-0.53320301,  0.22961526,  0.08092775],
       [ 0.21699124, -0.04954298, -0.02148484]])]

    def _compare_gradients(self, gx_numeric, gx_backward, directions):
        atol = self.atol
        rtol = self.rtol
        # Compare the gradients
        try:
            testing.assert_allclose(
                gx_numeric, gx_backward, atol=atol, rtol=rtol)
        except AssertionError as e:
            eps = self.eps
            x_data = self.x_data
            y_grad = self.y_grad
            f = six.StringIO()
            f.write('check_backward failed (eps={} atol={} rtol={})\n'.format(
                eps, atol, rtol))
            for i, x_ in enumerate(x_data):
                f.write('inputs[{}]:\n'.format(i))
                f.write('{}\n'.format(x_))
            for i, gy_ in enumerate(y_grad):
                f.write('grad_outputs[{}]:\n'.format(i))
                f.write('{}\n'.format(gy_))
            for i, d_ in enumerate(directions):
                f.write('directions[{}]:\n'.format(i))
                f.write('{}\n'.format(d_))
            f.write('gradients (numeric):  {}\n'.format(gx_numeric))
            f.write('gradients (backward): {}\n'.format(gx_backward))
            f.write('\n')
            f.write(str(e))
>           raise AssertionError(f.getvalue())
E           AssertionError: Parameterized test failed.
E           
E           Base test method: TestDeCov.test_backward_gpu
E           Test parameters:
E             backward_options: {'atol': 0.03}
E             dtype: <class 'numpy.float16'>
E             forward_options: {'rtol': 0.01, 'atol': 0.01}
E             reduce: no
E           
E           
E           (caused by)
E           AssertionError: check_backward failed (eps=0.02 atol=0.03 rtol=0.0001)
E           inputs[0]:
E           [[-0.70947266 -0.93554688 -0.78369141]
E            [-0.48706055  0.58935547  0.96044922]
E            [-0.33935547  0.55810547  0.56396484]
E            [ 0.88574219  0.99462891  0.06210327]]
E           grad_outputs[0]:
E           [[ 0.04333496 -0.74707031 -0.53125   ]
E            [-0.39941406  0.02464294 -0.60693359]
E            [-0.95996094 -0.88964844 -0.82714844]]
E           directions[0]:
E           [[-0.1604053   0.12723993  0.4412627 ]
E            [ 0.5194434   0.30622899  0.0786798 ]
E            [-0.53320301  0.22961526  0.08092775]
E            [ 0.21699124 -0.04954298 -0.02148484]]
E           gradients (numeric):  0.049424171447753906
E           gradients (backward): 0.08145464773508873
E           
E           
E           Not equal to tolerance rtol=0.0001, atol=0.03
E           
E           (mismatch 100.0%)
E            x: array(0.049424171447753906)
E            y: array(0.08145464773508873)
E           
E           assert_allclose failed: 
E             shape: () ()
E             dtype: float64 float64
E             i: (0,)
E             x[i]: 0.049424171447753906
E             y[i]: 0.08145464773508873
E             relative error[i]: 0.3932307999355187
E             absolute error[i]: 0.032030476287334825
E           x: 0.049424171447753906
E           y: 0.08145464773508873

atol       = 0.03
d_         = array([[-0.1604053 ,  0.12723993,  0.4412627 ],
       [ 0.5194434 ,  0.30622899,  0.0786798 ],
       [-0.53320301,  0.22961526,  0.08092775],
       [ 0.21699124, -0.04954298, -0.02148484]])
directions = [array([[-0.1604053 ,  0.12723993,  0.4412627 ],
       [ 0.5194434 ,  0.30622899,  0.0786798 ],
       [-0.53320301,  0.22961526,  0.08092775],
       [ 0.21699124, -0.04954298, -0.02148484]])]
eps        = 0.02
f          = <_io.StringIO object at 0x7f83d16e01f8>
gx_backward = array(0.08145464773508873)
gx_numeric = array(0.049424171447753906)
gy_        = array([[ 0.04333496, -0.74707031, -0.53125   ],
       [-0.39941406,  0.02464294, -0.60693359],
       [-0.95996094, -0.88964844, -0.82714844]], dtype=float16)
i          = 0
rtol       = 0.0001
self       = <chainer.gradient_check._CheckBackward object at 0x7f83d16deda0>
x_         = array([[-0.70947266, -0.93554688, -0.78369141],
       [-0.48706055,  0.58935547,  0.96044922],
       [-0.33935547,  0.55810547,  0.56396484],
       [ 0.88574219,  0.99462891,  0.06210327]], dtype=float16)
x_data     = (array([[-0.70947266, -0.93554688, -0.78369141],
       [-0.48706055,  0.58935547,  0.96044922],
       [-0.33935547,  0.55810547,  0.56396484],
       [ 0.88574219,  0.99462891,  0.06210327]], dtype=float16),)
y_grad     = (array([[ 0.04333496, -0.74707031, -0.53125   ],
       [-0.39941406,  0.02464294, -0.60693359],
       [-0.95996094, -0.88964844, -0.82714844]], dtype=float16),)
```

--- a/tests/chainer_tests/functions_tests/loss_tests/test_decov.py
+++ b/tests/chainer_tests/functions_tests/loss_tests/test_decov.py
@@ -12,7 +12,6 @@ from chainer.testing import backend
 
 
 def _decov(h):
-    h = cuda.to_cpu(h)
     h_mean = h.mean(axis=0)
     N, M = h.shape
     loss_expect = numpy.zeros((M, M), dtype=h.dtype)

--- a/tests/chainer_tests/functions_tests/loss_tests/test_decov.py
+++ b/tests/chainer_tests/functions_tests/loss_tests/test_decov.py
@@ -11,7 +11,7 @@ from chainer.testing import attr
 from chainer.testing import backend
 
 
-def _deconv(h):
+def _decov(h):
     h = cuda.to_cpu(h)
     h_mean = h.mean(axis=0)
     N, M = h.shape
@@ -25,15 +25,10 @@ def _deconv(h):
     return loss_expect / N
 
 
-@testing.parameterize(*testing.product_dict(
-    [{'dtype': numpy.float16},
-     {'dtype': numpy.float32},
-     {'dtype': numpy.float64},
-     ],
-    [{'reduce': 'half_squared_sum'},
-     {'reduce': 'no'},
-     ],
-))
+@testing.parameterize(*testing.product({
+    'dtype': [numpy.float16, numpy.float32, numpy.float64],
+    'reduce': ['half_squared_sum', 'no'],
+}))
 @backend.inject_backend_tests(
     None,
     # CPU tests
@@ -56,12 +51,12 @@ class TestDeCov(testing.FunctionTestCase):
             self.check_backward_options.update({'atol': 1e-3, 'eps': 0.02})
 
     def generate_inputs(self):
-        self.h = numpy.random.uniform(-1, 1, (4, 3)).astype(self.dtype)
-        return self.h,
+        h = numpy.random.uniform(-1, 1, (4, 3)).astype(self.dtype)
+        return h,
 
     def forward_expected(self, inputs):
-        h_data, = inputs
-        loss_expect = _deconv(h_data)
+        h, = inputs
+        loss_expect = _decov(h)
         if self.reduce == 'half_squared_sum':
             loss_expect = (loss_expect ** 2).sum() * 0.5
         return chainer.utils.force_array(loss_expect, self.dtype),

--- a/tests/chainer_tests/functions_tests/loss_tests/test_decov.py
+++ b/tests/chainer_tests/functions_tests/loss_tests/test_decov.py
@@ -6,9 +6,9 @@ import six
 import chainer
 from chainer.backends import cuda
 from chainer import functions
-from chainer import gradient_check
 from chainer import testing
 from chainer.testing import attr
+from chainer.testing import backend
 
 
 def _deconv(h):
@@ -26,84 +26,50 @@ def _deconv(h):
 
 
 @testing.parameterize(*testing.product_dict(
-    [{'dtype': numpy.float16,
-      'forward_options': {'rtol': 1e-2, 'atol': 1e-2},
-      'backward_options': {'atol': 3e-2}},
-     {'dtype': numpy.float32,
-      'forward_options': {'rtol': 1e-4, 'atol': 1e-4},
-      'backward_options': {'atol': 1e-3}},
-     {'dtype': numpy.float64,
-      'forward_options': {'rtol': 1e-4, 'atol': 1e-4},
-      'backward_options': {'atol': 1e-3}},
+    [{'dtype': numpy.float16},
+     {'dtype': numpy.float32},
+     {'dtype': numpy.float64},
      ],
     [{'reduce': 'half_squared_sum'},
      {'reduce': 'no'},
      ],
 ))
-class TestDeCov(unittest.TestCase):
+@backend.inject_backend_tests(
+    None,
+    # CPU tests
+    [{}]
+    # GPU tests
+    + [{
+        'use_cuda': True,
+    }]
+)
+class TestDeCov(testing.FunctionTestCase):
+
+    skip_double_backward_test = True
 
     def setUp(self):
-        self.h = numpy.random.uniform(-1, 1, (4, 3)).astype(self.dtype)
-        if self.reduce == 'half_squared_sum':
-            gloss_shape = ()
+        if self.dtype == numpy.float16:
+            self.check_forward_options.update({'rtol': 1e-2, 'atol': 1e-2})
+            self.check_backward_options.update({'atol': 3e-2, 'eps': 0.02})
         else:
-            gloss_shape = (3, 3)
-        self.gloss = numpy.random.uniform(
-            -1, 1, gloss_shape).astype(self.dtype)
+            self.check_forward_options.update({'rtol': 1e-4, 'atol': 1e-4})
+            self.check_backward_options.update({'atol': 1e-3, 'eps': 0.02})
 
-    def check_forward(self, h_data):
-        h = chainer.Variable(h_data)
-        loss = functions.decov(h, self.reduce)
-        self.assertEqual(loss.shape, self.gloss.shape)
-        self.assertEqual(loss.data.dtype, self.dtype)
-        loss_value = cuda.to_cpu(loss.data)
+    def generate_inputs(self):
+        self.h = numpy.random.uniform(-1, 1, (4, 3)).astype(self.dtype)
+        return self.h,
 
-        # Compute expected value
-        h_data = cuda.to_cpu(h_data)
-
+    def forward_expected(self, inputs):
+        h_data, = inputs
         loss_expect = _deconv(h_data)
         if self.reduce == 'half_squared_sum':
             loss_expect = (loss_expect ** 2).sum() * 0.5
+        return chainer.utils.force_array(loss_expect, self.dtype),
 
-        numpy.testing.assert_allclose(
-            loss_expect, loss_value, **self.forward_options)
-
-    def test_forward_cpu(self):
-        self.check_forward(self.h)
-
-    @attr.gpu
-    def test_forward_gpu(self):
-        self.check_forward(cuda.to_gpu(self.h))
-
-    def check_backward(self, h_data, gloss_data):
-        def f(h):
-            return functions.decov(h, self.reduce)
-
-        gradient_check.check_backward(
-            f, (h_data,), gloss_data, eps=0.02, **self.backward_options)
-
-    def check_type(self, h_data, gloss_data):
-        h = chainer.Variable(h_data)
+    def forward(self, inputs, device):
+        h, = inputs
         loss = functions.decov(h, self.reduce)
-        loss.grad = gloss_data
-        loss.backward()
-        self.assertEqual(h_data.dtype, h.grad.dtype)
-
-    def test_backward_cpu(self):
-        self.check_backward(self.h, self.gloss)
-
-    def test_backward_type_cpu(self):
-        self.check_type(self.h, self.gloss)
-
-    @attr.gpu
-    def test_backward_type_gpu(self):
-        self.check_type(cuda.to_gpu(self.h),
-                        cuda.to_gpu(self.gloss))
-
-    @attr.gpu
-    def test_backward_gpu(self):
-        self.check_backward(cuda.to_gpu(self.h),
-                            cuda.to_gpu(self.gloss))
+        return loss,
 
 
 class TestDeconvInvalidReductionOption(unittest.TestCase):

